### DISCUSSION TITLE
Visualisation tools for Godot

### DISCUSSION THREAD

**Discussion by ghost** [godot-extended-libraries/godot-ideas#30] [Ideas]
## Context
PyCharm and other JetBrains IDEs have this really useful "SciView" feature, which lets you look through arrays quite prettily. Image below:

![image](https://user-images.githubusercontent.com/24230522/65641668-4e8ed880-dfe5-11e9-9e31-5ac18469b06d.png)

The data visualisation window is on the top right, code in the middle. I'm trying to write a basic terrain generation script (i.e. plug values from the OpenSimplexNoise object into SurfaceTool and generate a MeshInstance from that). Using this kind of tool would've really helped a lot, just so I could've seen what values I'm getting easily, rather than printing it all out _like a savage_.

## Suggestion
Let's implement a feature like this! The way the JetBrains IDE family has implemented it is pretty good, and there's no immediate improvements that I think of, but if anyone can think of anything else useful, please say so.

I repeat, **if you feel like there could be other useful data visualisation tools to include, please say so**.

I'd like to try and work on it myself (alongside working on my other projects), but I wouldn't know where to start, frankly. Looking through the documentation, it was very confusing to tell the difference between a tool, module, plugin, or whatever other things there are. So, some advice on the best path to implement this would be nice as well! (Ideally, I'd write in such a way that it can easily be integrated into the main editor in a later release...)

Cheers,
    ~ Curradium

**Comment by willnationsdev** [comment]
This would be a plugin. See godotengine/godot-proposals#13 which is already a suggestion of having a tabular data editor window. I would just pipe in there and comment that you'd like a version that can deal with Pool*Arrays rather than just Resources (which makes sense).

A "tool" is anything that runs in the editor context (not at runtime).

A "module" is a collection of C++ classes that are registered under the `/modules` folder of the main engine repository. They are compiled directly into the engine binary.

An "addon" is a downloadable asset collection from online somewhere. Often times includes scripts, textures, models, etc.

A "plugin" is an addon that contains an EditorPlugin in the `/addons/<plugin_name>` directory. You can generate the code for a plugin by going to ProjectSettings, to the Plugins tab, and then clicking the "Create" button and filling out the info in the dialog. It will then also open the EditorPlugin script that is generated for you. Plugins make changes to Godot's Editor.

Modules are also capable of having C++ written EditorPlugins that modify the editor.

A "GDNative project/plugin/addon/module" (it's really just an addon) is any compiled GDNative code (C++, Rust, D, Haskell, whatever) that is dynamically linked to the engine code. They can run at engine time or editor time (if the language supports tool scripts - most do). They can be added outside of the compiled engine binary, so you can install them as an asset.

**Comment by ghost** [comment]
Crikey, I wish I could type that fast and well in 8 minutes!

Thanks for all your input @willnationsdev, I'll drop a message into that godot-proposals issue you linked soon; I appreciate all the info.
